aya/src/programs/extension.rs

```diff
@@ -37,12 +37,22 @@ pub enum ExtensionError {
 /// use aya::{BpfLoader, programs::{Xdp, XdpFlags, Extension}};
 ///
 /// let mut bpf = BpfLoader::new().extension("extension").load_file("app.o")?;
-/// let prog: &mut Xdp = bpf.program_mut("main").unwrap().try_into()?;
+/// let mut prog = None;
+/// let mut ext = None;
+/// for (name, program) in bpf.programs_mut() {
+///     match name {
+///         "main" => prog = Some(program),
+///         "extension" => ext = Some(program),
+///         _ => {},
+///     }
+/// }
+///
+/// let prog: &mut Xdp = prog.unwrap().try_into()?;
```

Comment:
Requiring to loop through the programs whenever we need to attach extensions doesn't seem ideal. I wonder what would the performance difference be of that versus having something like:

```rust
let extension = bpf.with_program_mut("extension", |bpf: &mut Bpf, extension: &mut Program| {
    let extensio: &mut Xdp = extension.try_into()?;
    let prog = bpf.program_mut("main").unwrap();
    prog.load()?;
    prog.attach("eth0", XdpFlags::default())?;
    extension.load(prog.fd().unwrap())?;
    Ok(extension)
}).unwrap()?;

extension.attach()?;
```

The way `with_program_mut` would be implemented would be that it'd essentially remove the given program from the program map, pass a mutable reference to the closure, and after the closure is finished put that program back in. The tricky thing I can see with this is that if the closure panics we aren't putting the program back in so it might get to an inconsistent state but I don't immediately see that as a big deal IMO.

Reply:
I don't think this scales all that well - in particular what happens if you want to operate on N extensions at the same time? In the fullness of time we should use https://github.com/rust-lang/rust/issues/97601 (get_many_mut) but that's not available today. Though it is available via hashbrown.

Maybe @alessandrod has ideas?

Reply:
Pulled in `hashbrown`. See how you like this.

Reply:
This is one of the reasons I said let's used OnwedFd and try_clone_to_owned in public APIs that expose FDs in #723 - as well as taking the same approach for Links.

I'm not opposed to optimizing later to use BorrowedFd, and even adding `get_many_mut`, but these changes probably require some design work upfront.
As it stands this PR will break users that are expecting the mutable borrow from `program_mut` to be dropped after getting a `ProgramFd` - aside from the example here this could also be an issue for users of `ProgramArray`.

We need to work up a design around how lifetimes in `BpfLoader` would work with all the myriad bpf lifecycle APIs we have.

Reply:
Agreed; deferring such design seems better. Done.

Reply:
I don't love the idea that the only way to get access to the file descriptor of a program by an user is by dup'ing it especially since it seems we are doing this to get around the awkwardness of passing a loaded program's fd to an extension which isn't the only reason one may need to get the file descriptor for. A use case I have for needing the file descriptor is for setting up the tail call table which doesn't require mutable references of multiple programs but with this change I now would have to pay the runtime cost created to work around extensions on a different workflow. 

Fwiw the example code I gave above would have worked for multiple extensions as well unless I am missing something:

```rust
// If you need to attach one program to many extensions
let extension = bpf.with_program_mut("main", |bpf: &mut Bpf, prog: &mut Program| {
   let prog: &mut Xdp = prog.try_into()?;
    prog.load()?;
    prog.attach("eth0", XdpFlags::default())?;
    let prog_fd = prog.fd().unwrap()?;
    
    let extension: &mut Extension = bpf.program_mut("extension1").unwrap().try_into()?;
    extension.load(prog_fd))?;
    extension.attach()?;
   
     // no longer need extension here so we can shadow the variable  and the borrow checker is happy
    let extension: &mut Extension = bpf.program_mut("extension2").unwrap().try_into()?;
    extension.load(prog_fd))?;
    extension.attach()?;
    
    Ok(())
}).unwrap()?;
```

The above *also* solves the issue of needing to borrow from the maps and programs at the same time fwiw which would be useful for the socketfd stuff.